### DISCUSSION TITLE
Testy integracyjne i End-to-End na Azure (DevOps)

### DISCUSSION THREAD

**Discussion by MichalKosowski** [AzureCommunityPL/group#47] [Pomoc] [answered]
Hej,

Mamy na Azure szybko rosnący system który skład się z dość szerokiej gamy komponentów, w skrócie EventHub zbiera eventy i zapisuje w CosmosDB a potem mamy różne mikroserwisy i EventGrid pomiędzy do komunikacji a dalej kolejne bazy i klientów. Dość standardowy obraz.

Deploy wszystkiego robimy przez Azure DevOps na którym kazdy serwis, czasem zbiór serwisów, ma osobny pipeline - a więc różne elementy mają deploy w różnym czasie.

Teraz problem - mamy unit testy, ale nie mamy testów integracyjnych a tym bardziej end to end. Co byśmy chcieli zrobic:
- test komponentu / serwisu na Azure, czy odpowiada i działa jak trzeba
- test komunikacji pomiędzy komponenetami
- test deploymentu resourców
- end-to-end test, czyli wrzucamy różne strumienie danych wejściowych i sprawdzamy czy doszły one do kazdego punktu i czy na końcu mamy wyniki które nas zadowalają (np. odpowiednio zagregowane dane pod widok).

Próbuje wyszukać jakiś przykładów i wzorców jak to zrobic najlepiej ale nie bardzo znajduje. To o czym myślę to:
- dla integracji, wysyłać spreparowane eventy na event grida pod dany serwis i odpytywać endpointy
- dla end to end, to samo, tyle że na większą skalę i większą ilość serwisów na raz.

Pytanie czy da się to zrobić lepiej? :) Acha - testy bez klientów UI - tylko backend stack!

Pozdrawiam,
Michał

**Comment by kamil-mrzyglod** [comment, marked as answer]
Pytanie kontrolne - jeśli robisz test E2E, który pokrywa wszystkie poprzednie scenariusze, to po co Ci oddzielne testy komponentów? Wymóg regulacyjny czy oddolna inicjatywa?

**Reply by MichalKosowski** [reply]
Hej,

Myślę że trochę inaczej. Założenie jest takie że zrobienie czegoś end-to-end może zająć dużo czasu a więc w teorii powinno być łatwiej zacząć od poszczególnych modułów a potem połączyć to w całość. Ale mogę się tutaj mylić. Wymogu nie ma, bardziej oddolna inicjatywa :)

**Reply by kamil-mrzyglod** [reply]
Pytam, nie oceniam :) Osobiście przy tego typu testach(czyli logika aplikacji na podstawie infry) to bym odpuścił sobie testowanie indywidualnych komponentów i skupił się na upewnieniu, że całość działa. Powodów jest wiele:
* infrastruktura chmurowa - tutaj zawsze rozważasz scenariusz, że coś nie działa
* architektura mikroserwisowa, gdzie jednostkowe testowanie nie odzwierciedla realnych scenariuszy
* logika oparta o zdarzenia a przy użyciu wylistowanych przez Ciebie komponentów nie ma gwarancji kolejności danych

Ja bym zeskryptował całą infrastrukturę i zakładając, że macie zestaw danych wg którego można badać zachowanie systemu, stawiał ją na potrzeby testów. Całościowo to pewnie nieco więcej pracy, nie jestem jednak pewien czy składanie tego z kilku elementów nie zajmie przypadkiem jeszcze więcej.

Czego używacie do czytania zdarzeń z Cosmosa? Change Feeda?

**Reply by MichalKosowski** [reply]
Nie ma problemu - każdy wartościowy komentarz mile widziany :)

Zaczynając od końca. Tak, mamy CosmosDB i ChangeFeed a potem Azure Functions które robią publish na EventGrid. To jest świerzy system i dużo się uczymy. Mamy też dość dużą swobodę rozwoju.

Właśnie nie jestem pewien czy da się wszystko zeskryptować - brakuje mi kilku elementów ciągle więc przy stawianiu nowego środowiska coś tam jeszcze ręcznie niestety ustawiamy (głównie: manage identity do serwisów aby był dostęp do keyvault i keyvault references do sekretów oraz subskrypcje na event grid). Zapewne da się to jakoś zautomatyzować, ale na razie odpuściliśmy temat :)

I tutaj też jest moje kolejne pytanie. Bo rozumiem można to robić na dwa sposoby:
- albo na potrzeby testów tworzysz nowe środowisko (cały Azure), robisz testy i potem wszystko usuwasz
- albo np. robisz deploy nowej wersji któregoś komponentu na dane środowisko i po deployu puszczasz testy (albo co wieczór).

Aktualnie wydaje mi się że lepiej zrobić to na istniejące środowisko a nie tworzone ad hoc, ale nie mam tutaj pewności...

**Reply by kamil-mrzyglod** [reply]
Ważne aby robić to konsekwentnie. Realnie najlepiej pewnie to połączyć:
* ćwiczysz stawianie infry od początku gdzieś z boku aby mieć pewność, że bez problemu stworzysz kopię gdyby była potrzeba
* robisz cykliczny deployment na istniejącą infrastrukturę

Managed Identity i resztę bez problemu zeskryptujesz. IMHO to jest podstawa i zapewne pokryje jeden z Twoich scenariuszy(deployment infrastruktury). 

Minus deploymentu na istniejące środowisko to brak gwarancji, że jesteś w stanie od zera postawić system. Z drugiej strony zwykle na proda robisz inkrementalny deployment więc tutaj trzeba też temat jakoś ugryźć.

**Reply by MichalKosowski** [reply]
Dzięki wielkie za komentarze. Postaram się wszystko co zrobiłem jakoś opisać i wrzucić tutaj lub na bloga - o ile coś ciekawego powstanie...

**Comment by MichalKosowski** [comment]
Acha, jako źródło danych rozważamy dwa scenariusze:
1) kontrolny, przygotowany zestaw danych, taki gdzie łatwo możemy ustalić szczegółowy output
2) zaciąganie losowych (ostatnich?) danych produkcyjnych i ponowne przepuszczanie ich przez system.

**Comment by MichalKosowski** [comment]
Hej @kamil-mrzyglod dzięki za dyskusję. Stworzyłem PoC flow i wydaje się że wszystko działa tak jak powinno. Muszę to jeszcze dopracować a więc pewnie za jakiś miesiąc lub więcej postaram się opisać (chyba że zmienię projekt 😅 ). Na ten moment zamykam dyskusje - problem rozwiązany.
